File: topics/mpi/practicals/02.MPI_pt2pt/fortran/2.ping_pong_solution.f90
PROGRAM ping_pong

!==============================================================!
!                                                              !
! This file has been written as a sample solution to an        !
! exercise in a course given at the CSCS Summer School.        !
! It is made freely available with the understanding that      !
! every copy of this file must include this header and that    !
! CSCS take no responsibility for the use of the enclosed      !
! teaching material.                                           !
!                                                              !
! Purpose: A program to try MPI_Ssend and MPI_Recv.            !
!                                                              !
! Contents: F-Source                                           !
!==============================================================!

  USE MPI
  IMPLICIT NONE

  INTEGER length
  PARAMETER (length=1)

  INTEGER PING
  PARAMETER (PING=0)

  INTEGER PONG
  PARAMETER (PONG=1)

  INTEGER status(MPI_STATUS_SIZE)

  REAL buffer(length)

  INTEGER ierror, my_rank

  CALL MPI_INIT(ierror)

  CALL MPI_COMM_RANK(MPI_COMM_WORLD, my_rank, ierror)

  ! Process 0 sends a message(ping) to process 1.
  ! After receiving the message, process 1 sends a message (pong) to process 0
  IF (my_rank.EQ.0) THEN
     CALL MPI_SEND(buffer, length, MPI_REAL, 1, PING, MPI_COMM_WORLD, ierror)
     CALL MPI_RECV(buffer, length, MPI_REAL, 1, PONG, MPI_COMM_WORLD, status, ierror)
  ELSE IF (my_rank.EQ.1) THEN
     CALL MPI_RECV(buffer, length, MPI_REAL, 0, PING, MPI_COMM_WORLD, status, ierror)
     CALL MPI_SEND(buffer, length, MPI_REAL, 0, PONG, MPI_COMM_WORLD, ierror)
  END IF

  WRITE (*,*) 'Ping-ping completed - rank ',my_rank

  CALL MPI_FINALIZE(ierror)

END PROGRAM

File: topics/mpi/practicals/02.MPI_pt2pt/cxx/2.ping_pong_solution.c
/****************************************************************
 *                                                              *
 * This file has been written as a sample solution to an        *
 * exercise in a course given at the CSCS Summer School.        *
 * It is made freely available with the understanding that      *
 * every copy of this file must include this header and that    *
 * CSCS take no responsibility for the use of the enclosed      *
 * teaching material.                                           *
 *                                                              *
 * Purpose: A ping-pong                                         *
 *                                                              *
 * Contents: C-Source                                           *
 *                                                              *
 ****************************************************************/

#include <stdio.h>
#include <mpi.h>

#define PING  0 //message tag
#define PONG  1 //message tag
#define SIZE  1024

int main(int argc, char *argv[])
{
    int my_rank;
    float buffer[SIZE];
    MPI_Status status;

    MPI_Init(&argc, &argv);

    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

    /* Process 0 sends a message (ping) to process 1.
     * After receiving the message, process 1 sends a message (pong) to process 0.
     */
    if (my_rank == 0) {
        MPI_Send(buffer, SIZE, MPI_FLOAT, 1, PING, MPI_COMM_WORLD);
        MPI_Recv(buffer, SIZE, MPI_FLOAT, 1, PONG, MPI_COMM_WORLD, &status);
    }
    if (my_rank == 1) {
        MPI_Recv(buffer, SIZE, MPI_FLOAT, 0, PING, MPI_COMM_WORLD, &status);
        MPI_Send(buffer, SIZE, MPI_FLOAT, 0, PONG, MPI_COMM_WORLD);
    }
    printf("Rank %d says: Ping-pong is completed.\n",my_rank);

    MPI_Finalize();
    return 0;
}
